staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go

```diff
@@ -238,15 +238,23 @@ const (
 // to resource directories enforcing namespace rules.
 func NoNamespaceKeyRootFunc(ctx context.Context, prefix string) string {
 	key := prefix
+	shard, err := genericapirequest.ValidShardFrom(ctx)
```

Comment:
we need a better way of plugging a shard name into the storage layer.
something that the kcp and the cache server could customize upon initialization.

Reply:
initially, I wanted to plug in different implementations but I realised the kcp and the cache server might share the same database instance.